### DISCUSSION TITLE
Table inside Tabs and Modal

### DISCUSSION THREAD

**Discussion by monointeractive** [filamentphp/filament#8293] [Help]
Hi. I would be asking for help.
I use standalone libraries like Table Builder and Form Builder version 3 (without Panel Builder).

1) How to put a table defined in public function table(){...} inside Tabs components.
2) How to put a table defined in public function table(){...} inside Modal components.

Thank you in advance for your help.

**Comment by leandrocfe** [comment]
What code did you try?
Not sure, but I think the table would work better on a separate page rather than in a modal.

**Reply by monointeractive** [reply]
1) I have no problem using a table as a livewire component.
2) I have no problem using tabs as Livewire component.
According to this instruction: https://filamentphp.com/docs/3.x/infolists/adding-an-infolist-to-a-livewire-component

I can even do it within a single component/page.

I'm having trouble placing the table from point 1 to the tabs (point 2). 
I've tried everything I could find in the documentation, so it would be hard to get a code snippet right now.

The same problem applies to tables in modals.
Is there anyone here who has succeeded?

**Reply by kucharskill** [reply]
I'm also very interested if anyone has managed to embed a Table in a modal window, or a table in a form, or some other InfoList component.

However, if this is not possible, the functionality of the Filament would be extremely limited and its use would limit the possibility of developing the application in which it is used (when a business need).

I hope someone has done it.

**Reply by danharrin** [reply]
You can embed a table using a `View::make()` component with a Livewire component inside, but you cannot use any actions or bulk actions with the nested table due to browser limitations

**Reply by Skullbock** [reply]
@danharrin i managed to create something that works, with actions within, by using the `modalContent()` method, a custom Livewire component with the table, and the `dispatch()` method of the table actions to interact with the parent. Would be cool to find a way to contribute a better way into Filament core if possible!

I find this technique super useful to allow selecting a related record (like we currently to in a Select `->relationship()` field) using a dedicated table with filters and everything. Especially useful if you need to search on hundreds or thousands 
of records, and the filter is not a simple text search but maybe based on categories, parameters, etc.

![Screenshot 2024-02-01 alle 14 43 11](https://github.com/filamentphp/filament/assets/1104083/fb4aac75-43b5-41b3-a2be-c076db97a130)

![Screenshot 2024-02-01 alle 14 43 17](https://github.com/filamentphp/filament/assets/1104083/33cd8012-2fb1-463b-b1ab-dcae3316dab6)

**Reply by photojo** [reply]
Can you share some snippets of your code?

**Reply by Skullbock** [reply]
The action to open the modal is

```php
Action::make('select')
      ->form([
            Hidden::make('motorPriceId')
        ])
        ->modalHeading(__('motor.motor'))
        ->modalContent(fn(Get $get) => new HtmlString(Blade::render('<div><livewire:motor-prices-table :fan-configuration="$fanConfiguration" /></div>', [
            'fanConfiguration' => $this->fanConfiguration($get)
        ])))
        ->action(fn(Set $set, array $data) => $set('motor.motorPriceId', $data['motorPriceId']))
```

and within the same component i have the listener for the event triggered by the table action:

```php
#[On('select-motor-price')]
    public function selectMotorPrice(int $record): void
    {
        $this
            ->getMountedFormComponentActionForm()
            ->fill([
                'motorPriceId' => $record
            ]);

        $this->callMountedFormComponentAction();
    }
```


The livewire component contains the table:

```php
public function table(Table $table): Table
    {
        return $table
            ->query(
                MotorPrice::query()
                    ->fromFanConfiguration($this->fanConfiguration)
                    ->orderBy('price')
            )
            ->columns([
                TextColumn::make('brand')
                    ->label(__('admin.brand')),
                TextColumn::make('model')
                    ->label(__('admin.model')),
                TextColumn::make('kW')
                    ->label(__('admin.kw')),
                TextColumn::make('efficiency')
                    ->label(__('admin.efficiency')),
                TextColumn::make('price')
                    ->label(__('admin.price')),
            ])
            ->filters([
                SelectFilter::make('brand')
                    ->label(__('admin.brand'))
                    ->options(
                        MotorPrice::query()
                            ->distinct('brand')
                            ->pluck('brand', 'brand')
                            ->toArray()
                    ),
                SelectFilter::make('efficiency')
                    ->label(__('admin.efficiency'))
                    ->options(Efficiency::class)

            ])
            ->actions([
                Action::make('select-motor-price')
                    ->label(__('admin.select'))
                    ->dispatch('select-motor-price', fn(MotorPrice $record) => [
                        'record' => $record->getKey()
                    ]),

            ]);
   }
```

**Reply by photojo** [reply]
I still have problems, but thank you.

**Reply by aidanwatt** [reply]
Hi @Skullbock your example code is exactly what I've been looking for, however when trying it out 

`$this->getMountedFormComponentActionForm()`

Returns null when the listening method (selectMotorPrice) is called, I've got this method within the custom component, is it meant to be elsewhere or am I missing something else? 

Thanks in advance

**Reply by Skullbock** [reply]
@aidanwatt the name of that method depends on where the action is. I am using a form action, but if you are not trying to using any other get mounted action method. There are several depending on where the action was placed.

**Reply by aidanwatt** [reply]
Thanks, I can't find a method that works, the action is a form component action:

```php
Forms\Components\Select::make('size_id')
                    ->relationship('size')
                    ->searchable()
                    ->searchLabels(true)
                    ->searchValues(false)
                    ->preload()
                    ->getOptionLabelFromRecordUsing(fn (Size $record) => "s: {$record->shape->name} w:{$record->width} h:{$record->length}")
                    ->required()
                    ->hintAction(
                        Forms\Components\Actions\Action::make('searchSizes')
                            ->icon('heroicon-m-magnifying-glass')
                            ->form([
                                Forms\Components\Hidden::make('size_id')
                            ])
                            ->modalWidth(MaxWidth::SevenExtraLarge)
                            //->modalSubmitAction(false)
                            ->modalCancelAction(false)
                            ->modalContent(fn() => new HtmlString(Blade::render('<div><livewire:sizes.search-sizes /></div>')))
                            ->action(fn(Forms\Set $set, array $data) => $set('size_id', $data['sizeId']))
                    ),
            ])

```

**Reply by Skullbock** [reply]
Then the form you're searching for is probably just the local form! `$this->form` or `$this->getForm('form-name')`

**Reply by aidanwatt** [reply]
Thanks, just realised I had the listener on the wrong component, just needed to be on the create page, all working with your original methods and making sense now. Many thanks

**Comment by monointeractive** [comment]
> @danharrin You can embed a table using a `View::make()` component with a Livewire component inside, but you cannot use any actions or bulk actions with the nested table due to browser limitations

Could you provide some simplest possible working code example?

**Reply by danharrin** [reply]
1) Create a Livewire component, add the table builder to it according to the docs
2) Create resources/views/filament/my-table.blade.php. Add in `@livewire(Path/To/Livewire/Table::class)`
3) In the form, `View::make('filament.my-table')`

**Reply by monointeractive** [reply]
@danharrin Thank you very much. Below is an example of a form class. Where can I add View::make... from point 3?


```
<?php
 use Livewire\Component;
[...] 
class ExampleForm extends Component implements HasForms {
    use InteractsWithForms;
    public function form(Form $form): Form
    {return $form->schema([...]);
    }
    
    public function render(): View{
     return <<<'HTML'
      {{ $this->form}}
     HTML;
    }
}
```

**Reply by leandrocfe** [reply]
```php
$form->schema([
View::make('filament.my-table')
]);
```
[view-components](https://filamentphp.com/docs/3.x/forms/layout/custom#view-components)

**Reply by moskoweb** [reply]
I managed to make the relationship appear in a table inside the modal, following @danharrin steps and adapting some elements:
Thanks @leandrocfe 

1. Create the Livewire Component
```
php artisan make:livewire-table HistoriesView Parent
```

2. In the generated HistoryView file, add the $record property so that it can be used to filter the table query.
```php
<?php

namespace App\Http\Livewire;

use Filament\Tables;
use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Tables\Contracts\HasTable;
use Illuminate\Contracts\View\View;
use Illuminate\Database\Eloquent\Builder;
use Livewire\Component;

class HistoriesView extends Component implements HasTable
{
    use InteractsWithTable;

    public $record;

    public function mount($record)
    {
        $this->record = $record;
    }
    
    protected function getTableQuery(): Builder
    {
        return Histories::query()
            ->where('model_id', $this->record->id);
    }

    protected function getTableColumns(): array
    {
        return [
            Tables\Columns\TextColumn::make('name'),
            Tables\Columns\TextColumn::make('status'),
        ];
    }

    public function render(): View
    {
        return view('livewire.histories-view');
    }
}
```

3. Create a view that will be added to the Livewire component and also pass the $record.
```
@livewire(\App\Http\Livewire\HistoriesView::class, ['record' => $record])
```

4. When calling the view in the Action, pass the record along with the view.
```php
Tables\Actions\Action::make('histories')
    ->action(fn () => $this->record)
    ->modalContent(
        fn ($record) => view('filament.widgets.details', [
            'record' => $record
        ])
    ),
```

**Reply by monointeractive** [reply]
Thanks @moskoweb. 
I think you used Filament version 2 not 3. Will it work the same in version 3?

**Reply by moskoweb** [reply]
Yes, I'm still using 2 in my project, because it's big I haven't migrated yet, but I believe that conceptually, it should be the same.

**Reply by popdaul** [reply]
thx dude～ @moskoweb

**Reply by RejownAhmed** [reply]
You are the best. Thanks dude @moskoweb

**Comment by photojo** [comment]
I'm using Filament 3. 
In my table I use

 ```
 Action::make('Edit Registrations')
                   ->modalContent(fn (Action $action, Registration $record): View => view(
                        'filament.pages.actions.advance',
                        ['action' => $action, 'record' => $record],
                    ))->slideOver()->modalSubmitAction(false)->modalCancelAction(false),

```

In 'advance' the form-component is rendered correctly.
`@livewire('MyForm', ['id' => $record->id])`

But the changes are not saved. Why? 
(Of course, the form 'MyForm' saves data, if called f.e. via route)

**Reply by leandrocfe** [reply]
Why not a modal form?
https://filamentphp.com/docs/3.x/actions/modals#modal-forms

**Comment by photojo** [comment]
Of course, a model is obvious. But I want to achieve the following:
One modal with several tabs, BUT each tab should have its own form.

**Reply by realihorrud** [reply]
Any solution?

**Reply by moskoweb** [reply]
These days I was creating a custom table inside a Widget, which I believe would be logical for you. What I needed to do was create a FakeModel, so that I could pass the necessary Query to the Table to work and render the information.

I took an Array of data, made a Collection with the Filament Model, and that's how it worked in the table.
